cli/serve/serve.go

```diff
@@ -244,9 +248,55 @@ func serverMain(args []string, c cli.Config) error {
 		log.Info("Now listening on ", addr)
 		return http.ListenAndServe(addr, nil)
 	}
+	if conf.MutualTLSCAFile != "" {
+		mutualTLSCAs, err := ioutil.ReadFile(conf.MutualTLSCAFile)
+		if err != nil {
+			return fmt.Errorf("Failed to load mutual TLS CA file: %s", err)
+		}
+		cert, err := ioutil.ReadFile(conf.TLSCertFile)
+		if err != nil {
+			return fmt.Errorf("Failed to load cert file: %s", err)
+		}
+		key, err := ioutil.ReadFile(conf.TLSKeyFile)
+		if err != nil {
+			return fmt.Errorf("Failed to load cert file: %s", err)
+		}
+		keypair, err := tls.X509KeyPair(cert, key)
+		if err != nil {
+			return fmt.Errorf("Malformed sever TLS material: %s", err)
+		}
 
-	log.Info("Now listening on https://", addr)
+		clientPool := x509.NewCertPool()
+		clientPool.AppendCertsFromPEM(mutualTLSCAs)
+		server := http.Server{
+			Addr: addr,
+			TLSConfig: &tls.Config{
+				Certificates: []tls.Certificate{keypair},
+				ClientAuth:   tls.RequireAndVerifyClientCert,
+				ClientCAs:    clientPool,
+			},
+		}
 
+		if conf.MutualTLSCN != "" {
+			log.Infof("Requiring CN=%s for client connections", conf.MutualTLSCN)
+			server.Handler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				if r != nil && r.TLS != nil {
+					connState := r.TLS
+					for _, peerCert := range connState.PeerCertificates {
+						if peerCert.Subject.CommonName == conf.MutualTLSCN {
```

Comment:
can we make it a regex filter? That would potentially make the whitelist more general.